File: app/src/main/java/com/example/spring/CreatePlanService.java
package com.example.spring;

import com.example.spring.error.ConflictException;
import com.example.spring.error.NotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Create a subscription plan that can later grant feature entitlements.
 */
@Service
public class CreatePlanService {

    private final EntitlementsRepository entitlementsRepository;

    public CreatePlanService(EntitlementsRepository entitlementsRepository) {
        this.entitlementsRepository = entitlementsRepository;
    }

    @Transactional
    public Plan handle(CreatePlanRequest request) {
        // spec-driven body for CreatePlan.
        // intent: Create a subscription plan that can later grant feature entitlements.
        // requires: EntitlementsRepository
        // return: Plan (entity); spec-driven body constructs from request inputs + persists via EntitlementsRepository.savePlan.
        Plan value = new Plan(request.tenantId(), request.planId(), request.planCode(), request.displayName(), request.billingInterval(), request.status(), request.createdAt());
        return this.entitlementsRepository.savePlan(value);
    }
}

File: app-python/entitlements_app/services/create_plan_service.py
"""Create a subscription plan that can later grant feature entitlements."""

from decimal import Decimal
import re

from fastapi import HTTPException, status

from entitlements_app.requests.create_plan_request import CreatePlanRequest
from entitlements_app.entities import Plan
from entitlements_app.ports import EntitlementsRepository


class CreatePlanService:
    """Service handling Command `CreatePlan`."""

    def __init__(self, entitlements_repository: EntitlementsRepository) -> None:
        self.entitlements_repository: EntitlementsRepository = entitlements_repository

    def handle(self, request: CreatePlanRequest) -> Plan:
        # spec-driven body for Command CreatePlan.
        # intent: Create a subscription plan that can later grant feature entitlements.
        # requires: EntitlementsRepository
        # return: Plan (entity); construct from request inputs and persist via EntitlementsRepository.savePlan.
        value = Plan(tenantId=request.tenantId, planId=request.planId, planCode=request.planCode, displayName=request.displayName, billingInterval=request.billingInterval, status=request.status, createdAt=request.createdAt)
        return self.entitlements_repository.savePlan(value)
